integration/experiment/ffnet/ffnet.py

```diff
@@ -0,0 +1,127 @@
+    if cpu_nn_path is None and gpu_nn_path is None:
+        raise RuntimeError('Must specify cpu-nn-path and/or gpu-nn-path when running ffnet experiment')
+
+    if cpu_nn_path is not None:
+        if os.path.exists(cpu_nn_path):
+            os.environ['GEOPM_CPU_NN_PATH'] = cpu_nn_path
+        else:
+            raise KeyError(f'File cpu-nn-path={cpu_nn_path} does not exist.')
+        if cpu_fmap_path is not None:
+            os.environ['GEOPM_CPU_FMAP_PATH'] = cpu_fmap_path
+        else:
+            raise RuntimeError('Must specify cpu-fmap-path when cpu-nn-path is specified for ffnet experiment')
+
+    if gpu_nn_path is not None:
+        if os.path.exists(gpu_nn_path):
+            os.environ['GEOPM_GPU_NN_PATH'] = gpu_nn_path
+        else:
+            raise KeyError(f'File gpu-nn-path={gpu_nn_path} does not exist.')
+        if gpu_fmap_path is not None:
+            os.environ['GEOPM_GPU_FMAP_PATH'] = gpu_fmap_path
+        else:
+            raise RuntimeError('Must specify gpu-fmap-path when gpu-nn-path is specified for ffnet experiment')
+
+    return
+
+def launch_configs(output_dir, app_conf, phi=0):
+    mach = machine.init_output_dir(output_dir)
+
+    if phi > 1 or phi < 0:
+        raise KeyError('perf-energy-bias must be between 0 and 1 for ffnet experiment.')
+
+    agent = 'ffnet'
+    targets = []
+    options = {"PERF_ENERGY_BIAS": phi}
+    name = f'{phi}phi'
+
+    file_name = os.path.join(output_dir, f'{agent}_agent_{name}.config'.format(agent))
+    agent_conf = geopmpy.agent.AgentConf(file_name, agent, options)
+    targets.append(launch_util.LaunchConfig(app_conf=app_conf,
+                                            agent_conf=agent_conf,
+                                            name=name))
+    return targets
+
+def launch(app_conf, args, experiment_cli_args):
+    output_dir = os.path.abspath(args.output_dir)
+    extra_cli_args = launch_util.geopm_signal_args(report_signals=report_signals(),
+                                                    trace_signals=trace_signals())
+    extra_cli_args += experiment_cli_args
+
+    setup_env_paths(args.cpu_nn_path,
+                    args.cpu_freq_rec_path,
+                    args.gpu_nn_path,
+                    args.gpu_freq_rec_path)
+
+    targets = launch_configs(output_dir, app_conf, args.perf_energy_bias)
+
+    #Set and initialize required counters for nn training
+    init_control_path = 'neural_net_init.controls'
+    with open(init_control_path, 'w') as outfile:
+        outfile.write(f"""MSR::PQR_ASSOC:RMID board 0 0
+                      # Assigns all cores to resource monitoring association ID 0
+                      # Next, assign resource monitoring ID for QM events to match
+                      f"MSR::QM_EVTSEL:RMID board 0 0
+                      # Then determine Xeon Uncore Utilization
+                      f"MSR::QM_EVTSEL:EVENT_ID board 0 0""")
```

Comment:
```suggestion
        outfile.write("""MSR::PQR_ASSOC:RMID board 0 0
                      # Assigns all cores to resource monitoring association ID 0
                      # Next, assign resource monitoring ID for QM events to match
                      MSR::QM_EVTSEL:RMID board 0 0
                      # Then determine Xeon Uncore Utilization
                      MSR::QM_EVTSEL:EVENT_ID board 0 0""")
```

Can remove f since there's no string interpolation here. Remove some quotation marks that are getting passed through to the outfile.